### PR TITLE
Reset embedded dirty checking

### PR DESCRIPTION
**Pull request by puneetbehl**
Updated GrailsEntityDirtinessStrategy.groovy to also reset dirty embedded objects.

Fixes grails/grails-data-mapping#1272

### CODE REVIEW

File: grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/dirty/GrailsEntityDirtinessStrategy.groovy [hunk 1]
```diff
@@ -56,6 +60,30 @@ class GrailsEntityDirtinessStrategy implements CustomEntityDirtinessStrategy {
...
+    private void resetDirtyEmbeddedObjects(PersistentEntity persistentEntity,
+                                           Object entity,
+                                           EntityPersister persister,
+                                           Session session) {
+
+        if (DirtyCheckingSupport.areEmbeddedDirty(persistentEntity, entity)) {
+            final associations = persistentEntity.getEmbedded()
+            for (Embedded a in associations) {
+                final value = a.reader.read(entity)
+                resetDirty(value, persister, session)
```
Comment (graemerocher) [inline grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/dirty/GrailsEntityDirtinessStrategy.groovy]: What if `value` is null?

File: grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/dirty/GrailsEntityDirtinessStrategy.groovy [hunk 2]
```diff
@@ -56,6 +60,30 @@ class GrailsEntityDirtinessStrategy implements CustomEntityDirtinessStrategy {
...
         if (canDirtyCheck(entity, persister, session)) {
             cast(entity).trackChanges()
+            try {
+                PersistentEntity persistentEntity = GormEnhancer.findEntity(Hibernate.getClass(entity))
+                if (persistentEntity != null) {
+                    resetDirtyEmbeddedObjects(persistentEntity, entity, persister, session)
+                }
+            } catch (IllegalStateException e) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug(e.message, e)
```
Comment (graemerocher) [inline grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/dirty/GrailsEntityDirtinessStrategy.groovy]: Why are we swallowing the exception here and logging at `debug` level?

File: grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/dirty/GrailsEntityDirtinessStrategy.groovy (hunk 1 above)
Comment (puneetbehl) [reply grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/dirty/GrailsEntityDirtinessStrategy.groovy]: If the value is `null` then we do not need to reset `$changedProperties` on embedded. So, the check at line#81 `DirtyCheckingSupport.areEmbeddedDirty(persistentEntity, entity)` would return `false`.

File: grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/dirty/GrailsEntityDirtinessStrategy.groovy (hunk 2 above)
Comment (puneetbehl) [reply grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/dirty/GrailsEntityDirtinessStrategy.groovy]: With this PR, we are recursively resetting the `$changedProperties` of embedded objects. In cases where the embedded object is not a PersistentEntity the recursion will break, and `IllegalStateException` is thrown as: 
```
Either class [grails.gorm.tests.dirtychecking.Address] is not a domain class or GORM has not been initialized correctly or has already been shutdown. Ensure GORM is loaded and configured correctly before calling any methods on a GORM entity.
```

**Comment by dbaylerg** [comment]
@graemerocher Is there are possibility to get this fix merged into the 6.1.x branch and create a 6.1.13 Release?  I am a month or two out from a major system replacement which my team has been working on and we have determined that this issue has become a significant problem as we have begun our Production Readiness Testing with large volumes of data.  

We used embedded objects throughout our domains.  So this issue manifests itself when we discover records with "Version" numbers in the hundreds when they should be 0,1 or 2.

The reason for this is because we batch process lots of these records.  For each batch a record is only updated once, but because of this issue, on the next batch the prior batch records are updated again.  On the next batch the prior 2 batches are updated again.  This results in hundreds of thousands of extra UPDATES to the database as we process larger number of batches.

Eventually the application grinds to a halt as the number of UPDATES cannot be handled.  I do not have the option to upgrade to GORM 7 at this time unfortunately.

**Pull request merged**